src/trait_manager/model/Traits.ts

```diff
@@ -33,30 +37,25 @@ export default class Traits extends Collection<Trait> {
     this.target = target;
   }
 
-  /** @ts-ignore */
-  add(models: string | Trait | TraitProperties | (string | Trait | TraitProperties)[], opt?: AddOptions) {
-    const em = this.em;
-
-    // Use TraitFactory if necessary
-    if (isString(models) || isArray(models)) {
-      const tm = em && em.get! && em.Traits;
-      const tmOpts = tm && tm.getConfig();
-      const tf = TraitFactory(tmOpts);
-
-      if (isString(models)) {
-        models = [models];
-      }
+  add(model: string | TraitProperties | Trait, options?: AddOptions): Trait;
+  add(models: Array<string | TraitProperties | Trait>, options?: AddOptions): Trait[];
+  add(models: unknown, opt?: AddOptions): any {
+    if (models == undefined) {
+      return undefined;
+    }
 
+    if (isArray(models)) {
+      var traits: Trait[] = [];
       for (var i = 0, len = models.length; i < len; i++) {
-        const str = models[i];
-        const model = isString(str) ? tf.build(str)[0] : str;
-        model.target = this.target;
-        models[i] = model as Trait;
+        const trait = models[i];
+        traits[i] = trait instanceof Trait ? trait : this.tf.build(trait);
+        traits[i].target = this.target;
       }
+      return super.add(traits, opt);
     }
+    const trait = models instanceof Trait ? models : this.tf.build(models as any);
+    trait.target = this.target;
 
-    return Collection.prototype.add.apply(this, [models as Trait[], opt]);
+    return super.add(trait, opt);
```

Comment:
I'm not sure exactly what is happening here but with these changes to Traits, undo/redo on traits doesn't work properly (eg. I tried my latest fix [here](https://github.com/GrapesJS/grapesjs/issues/4972) and it works only if I rollback these changes)

Reply:
I added a test to cover the undo scenario and fixed it :) Sorry I couldn't have time earlier  :(